filtering meals in kmm

Gmiller290488 opened the pull request:
I added code for Date in the shared code because all the libraries for it were causing me problems and I didn't need that much.
The filtering now happens in shared area.

Gmiller290488 committed:
Filterting happening in shared code.
Added shared code for Date and implementation for both Android and ios.
Added assertion error for api error.
Added no meals text when no meals are returned from api

Gmiller290488 committed:
Removed unneeded filtering code from ios

Gmiller290488 commented on iosHackthonApp/Modules/Features/MealList/Sources/MealList/MealListView.swift:
```diff
@@ -14,58 +14,64 @@ public struct MealListView: View {
+                        }
                     }
+                } else {
+                    Text("No meals found") // TODO: Do something nicer than this
```
Just a temporary thing so we can differentiate easier between no meals and problems with the API. We should make a nicer ui in the future for this scenario probably

Gmiller290488 commented on iosHackthonApp/Modules/Features/MealList/Sources/MealList/MealListViewModel.swift:
```diff
@@ -40,9 +40,11 @@ public final class MealListViewModel: ObservableObject {
                 let meals = meals,
-                error == nil else { return }
+                error == nil else {
+                assertionFailure(String(describing: error?.localizedDescription))
```
Same as above really. Just helpful for knowing whether there's no meals or a problem

KaneCheshire replied:
I think it's okay to force unwrap the error here to make this more readable since you've already checked that it's not nil:

`assertionFailure(error!.localizedDescription)` or `assertionFailure(error?.localizedDescription ?? "")`

Gmiller290488 commented on shared/src/androidMain/kotlin/com/kcc/kmmhackathon/shared/utility/DateKMM.kt:
```diff
@@ -0,0 +1,5 @@
+
+import java.util.*
+
+actual typealias DateKMM = Date
```
This java Date class fully matches the `expect` so we don't need to add anything else 🎉

KaneCheshire replied:
Nice!

Gmiller290488 commented on shared/src/commonMain/kotlin/com/kcc/kmmhackathon/shared/network/MealApi.kt:
```diff
@@ -5,6 +5,7 @@ import com.kcc.kmmhackathon.shared.entity.MealWithDistance
 import com.kcc.kmmhackathon.shared.entity.Quantity
 import com.kcc.kmmhackathon.shared.utility.DistanceUnit
 import com.kcc.kmmhackathon.shared.utility.LocationUtil
+import com.kcc.kmmhackathon.shared.utility.DateKMM
```
Had to call it DateKMM because Xcode gets itself all confused when I called it Date - which is to be expected, feel free to suggest a better name 🤷

KaneCheshire replied:
You can disambiguate by including the module name before when using Date like `Foundation.Date` or `Swift.Date` (can't remember which) and `shared.Date` but that might get tedious, so makes sense to typealias. Not super keen on the name, but it works. Maybe something like SharedDate, XDate (as in, cross platform date)

Gmiller290488 replied:
I might go with sharedDate 👍

Gmiller290488 commented on shared/src/commonMain/kotlin/com/kcc/kmmhackathon/shared/network/MealApi.kt:
```diff
@@ -45,11 +46,16 @@ class MealApi {
-        // TODO: Filter meals that have expired here
-        return meals.sortedWith(compareBy { it.distance })
+        return meals
+            .filterNot { DateKMM( it.expiryDate.toLong()).compareTo(DateKMM()) < 0 }
```
If the expiryDate is before the current date, -1 is returned

Gmiller290488 commented on shared/src/commonMain/kotlin/com/kcc/kmmhackathon/shared/utility/DateKMM.kt:
```diff
@@ -0,0 +1,18 @@
+
+    constructor(time: Long)
+
+    /**
+     * Compare this instance with the other instance.
+     *
+     * @param other: Another date instance.
+     * @return 0 if the other is equal to this instance; a value less than 0 if this instance is before the other;
+     *          and a value greater than 0 if this instance is after the other.
+     */
+    fun compareTo(other: DateKMM): Int
```
This is the documentation for the Java data compareTo function.
It's not the easiest function to understand, we could add our own compare function that returns a bool but I think then it would take a custom implementation of Date on Android side - but happy to do it if everyone thinks this is a bit gross

KaneCheshire replied:
I think there are some comparison APIs on the iOS side that return an enum (basically an Int) so I think it's okay to leave, might be nice to wrap up into something a bit more Swifty but fine as-is too :)

Gmiller290488 replied:
I extended ShareDate to have `isBefore`, `isAfter` and `isEqual` functions to make this a bit easier

Gmiller290488 committed:
Merge branch 'main' into chore/gm_filteringMealsInKMM

Gmiller290488 committed:
Indentation

aabolade commented on shared/src/iosMain/kotlin/com/kcc/kmmhackathon/shared/utility/DateKMM.kt:
```diff
@@ -0,0 +1,32 @@
+        date = platformDate
+    }
+
+    actual fun compareTo(other: DateKMM): Int {
```
Is there any way we can add tests for this logic?

Gmiller290488 replied:
Yeah currently doing it 👍

Gmiller290488 replied:
Done

Gmiller290488 committed:
Added tests for DateKMM

Gmiller290488 committed:
Merge branch 'main' into chore/gm_filteringMealsInKMM

KaneCheshire commented on iosHackthonApp/Modules/Features/MealList/Sources/MealList/MealListView.swift:
```diff
@@ -14,58 +14,64 @@ public struct MealListView: View {
-                        ForEach((0 ..< viewModel.meals.count), id: \.self) {
-                            MealRow(listViewModel: viewModel, rowViewModel: MealRowViewModel(meal: viewModel.meals[$0]))
-                                .padding()
+                if viewModel.meals.count > 0 {
```
How about
```swift
if viewModels.meals.isEmpty {
  Text("No meals found")
} else {
 ... etc
}
```

Gmiller290488 committed:
Addressing PR comments:
Changed DateKMM name to SharedDate.
Switched the if statement for showing no meals.

KaneCheshire commented on iosHackthonApp/Modules/Features/MealList/Sources/MealList/MealListViewModel.swift:
```diff
@@ -40,9 +40,11 @@ public final class MealListViewModel: ObservableObject {
                 let meals = meals,
-                error == nil else { return }
+                error == nil else {
+                assertionFailure(String(describing: error!.localizedDescription))
```
You don't need the `String(describing` bit now since that was just to fix the optional :)

Gmiller290488 replied:
Dopey of me 🤦

dsmith42 commented:
Looks good, really great to be lifting more of the logic up into the shared layer 🚀

dsmith42 reviewed (approved):
Looks good, really great to be lifting more logic into the shared layer.  Sharing is caring 🥳

Gmiller290488 committed:
Added extension to SharedDate.
Moved extensions to own folder.
Added tests for SharedDate extensions.

Gmiller290488 commented on shared/src/commonMain/kotlin/com/kcc/kmmhackathon/shared/utility/extensions/SharedDate+Extensions.kt:
```diff
@@ -0,0 +1,15 @@
+
+import com.kcc.kmmhackathon.shared.utility.SharedDate
+
+fun SharedDate.isAfter(other: SharedDate): Boolean {
+    return this.compareTo(other) > 0
+}
+
+fun SharedDate.isBefore(other: SharedDate): Boolean {
+    return this.compareTo(other) < 0
+}
+
+fun SharedDate.isEqual(other: SharedDate): Boolean {
+    return this.compareTo(other) == 0
+}
```
Added these because every time I looked at the `compareTo` code in the API file it made me go 🤯  and I was the one who wrote it

Gmiller290488 committed:
Removed excess brackets

Gmiller290488 commented on shared/src/commonMain/kotlin/com/kcc/kmmhackathon/shared/network/MealApi.kt:
```diff
@@ -45,11 +44,16 @@ class MealApi {
-        // TODO: Filter meals that have expired here
-        return meals.sortedWith(compareBy { it.distance })
+        return meals
+            .filter { SharedDate().isBefore(SharedDate(it.expiryDate.toLong())) }
```
filter meals where the date now, is before the expiry date.

I think this is easier to understand now

Gmiller290488 committed:
Merge branch 'main' into chore/gm_filteringMealsInKMM

Gmiller290488 committed:
Moved check for no mealse
